internal/controller/apiextensions/composite/api.go

```diff
@@ -358,17 +375,23 @@ type APIConfigurator struct {
 
 // Configure any required fields that were omitted from the composite resource
 // by copying them from its composition.
-func (c *APIConfigurator) Configure(ctx context.Context, cp xresource.Composite, rev *v1.CompositionRevision) error {
-	apiVersion, kind := cp.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
+func (c *APIConfigurator) Configure(ctx context.Context, cp resource.Composite, rev *v1.CompositionRevision) error {
+	// Only legacy XRs support writing connection secrets.
+	lcp, ok := cp.(resource.LegacyComposite)
+	if !ok {
+		return nil
+	}
```

Comment:
Calling attention to this since it's not boilerplate find/replace like most of the PR.